### PR TITLE
PyTorchDistributed use `group` as parameter instead of device

### PR DESCRIPTION
**Pull request by reyoung**
## Motivation

Make PyTorch Distributed integration support GPU distributed data-parallel.

## Description of the changes

Tensors must be on a CUDA device when using NCCL as a process group. So when using multiple GPU distributed data parallel, `optuna.TorchDistributedTrial` will throw an exception.

Createing a global `gloo` process group when WORLD is `NCCL` backend and carrying the `_g_pg` for all dist calls.

### CODE REVIEW

**Comment by reyoung** [comment]
@toshihikoyanase  Please take a review, thanks

**Comment by toshihikoyanase** [comment]
@reyoung Thank you for your PR. Before diving into details of code, could you share the minimum reproducible code of the problem? I guess we need it to confirm your change resolves the problem.

> Tensors must be on a CUDA device when using NCCL as a process group. So when using multiple GPU distributed data parallel, optuna.TorchDistributedTrial will throw an exception.

> Createing a global gloo process group

gloo seems to be used for the distributed CPU training. So, in my understanding, this PR uses the gloo backend for Optuna communication while it use NCCL for the communication of PyTorch's model training. Could you tell me if I'm wrong.

> Use the `Gloo` backend for distributed CPU training.
https://pytorch.org/docs/stable/distributed.html#which-backend-to-use

**Comment by toshihikoyanase** [comment]
I checked the implementation of `TorchDistriubtedTrial`. It moves data to devices (e.g., GPU memory) before communication and move it from the devices after communication. So, I think `TorchDistribuedTrial` is design to use distributed GPU training.
What do you think?

https://github.com/optuna/optuna/blob/2b8befe6fda8ca1290c1061b390ec2945377f530/optuna/integration/pytorch_distributed.py#L293-L294
https://github.com/optuna/optuna/blob/2b8befe6fda8ca1290c1061b390ec2945377f530/optuna/integration/pytorch_distributed.py#L300-L301

**Comment by reyoung** [comment]
I see.  It seems that the `device` need to be passed to `TorchDistributedTrial`.

However, moving tensors between CPU and GPU seems unnecessary and slow, and it is easier just to create a `gloo` backend when the WORLD is `NCCL`.

Also `device` is not needed because we can use `dist.get_backend` to detect the global process group's backend. When it is `NCCL`, just set `device` to `cuda:0`?

**Comment by reyoung** [comment]
@toshihikoyanase Please take a review.

**Comment by HideakiImamura** [comment]
@toshihikoyanase Could you review this PR?

File: optuna/integration/pytorch_distributed.py [hunk 1]
```diff
@@ -93,9 +94,13 @@ class TorchDistributedTrial(optuna.trial.BaseTrial):
         trial:
             A :class:`~optuna.trial.Trial` object or :obj:`None`. Please set trial object in
             rank-0 node and set :obj:`None` in the other rank node.
-        device:
```
Comment (HideakiImamura) [inline optuna/integration/pytorch_distributed.py:96]: The `device` argument is removed and this is distractive change. Do we need to consider the impact on users?

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR. I haven't really thought too much about the removal of the `device` argument and the addition of the `group` argument yet, but I'll leave a couple of quick comments first.

I would like to test the validity of this change. Is it possible to add unit tests? Note that Optuna does not test using gpu. optuna-examples confirms that TorchDistributedTrial works with cpu [here](https://github.com/optuna/optuna-examples/blob/main/pytorch/pytorch_distributed_simple.py). If possible, could you please provide a script to easily check the operation with Google colab or something?

File: optuna/integration/pytorch_distributed.py [hunk 2]
```diff
@@ -21,18 +22,18 @@
...
 
     _T = TypeVar("_T")
     _P = ParamSpec("_P")
 
-
 _suggest_deprecated_msg = (
     "Use :func:`~optuna.integration.TorchDistributedTrial.suggest_float` instead."
 )
 
+_g_pg: List[Optional["torch.distributed.ProcessGroup"]] = [None]
```
[the remark this answers is not in the document]
Comment (reyoung) [reply optuna/integration/pytorch_distributed.py:35]: We cannot.

Since the `TorchDistributedTrial` is created multiple times during training, we need a global variable to store `gloo` backend by default. Because process group is an application scoped variable, and `TorchDistributedTrial` is per-trial (or objective local ) variable.

See  https://github.com/optuna/optuna-examples/blob/main/pytorch/pytorch_distributed_simple.py#L100

File: optuna/integration/pytorch_distributed.py [hunk 3]
```diff
@@ -105,12 +110,30 @@ class TorchDistributedTrial(optuna.trial.BaseTrial):
...
 
-        if dist.get_rank() == 0:  # type: ignore
+        if group is not None:
+            self._group: "torch.distributed.ProcessGroup" = group
+        else:
+            if _g_pg[0] is None:
+                if dist.group.WORLD is None:
+                    raise RuntimeError("torch distributed is not initialized.")
+                default_pg: "torch.distributed.ProcessGroup" = dist.group.WORLD
+                if dist.get_backend(default_pg) == "nccl":  # type: ignore
```
[the remark this answers is not in the document]
Comment (reyoung) [reply optuna/integration/pytorch_distributed.py]: Fixed.

File: optuna/integration/pytorch_distributed.py [hunk 4]
```diff
@@ -21,18 +22,18 @@
     import torch
     import torch.distributed as dist
 
-
```
[the remark this answers is not in the document]
Comment (reyoung) [reply optuna/integration/pytorch_distributed.py]: Fixed.

File: optuna/integration/pytorch_distributed.py (hunk 1 above)
Comment (reyoung) [reply optuna/integration/pytorch_distributed.py:96]: Add an optional device argument. If device is not None, print warnings.

**Comment by reyoung** [comment]
For testing code,

Actually I changed the `huggingface/transformers` by using this PR .

[Code here](https://github.com/huggingface/transformers/compare/main...reyoung:transformers:feature/better_optuna_search)

I tested in my internal GPU cluster and it works well in multiple GPUs DDP.

However, writing unittests need a gpu CI system, which optuna does not have one?

**Comment by reyoung** [comment]
@HideakiImamura 

For example please see https://github.com/optuna/optuna-examples/pull/145.

Actually, the PyTorch distributed environments should not be set manually. You can use `torchrun` to set them all.

File: optuna/integration/pytorch_distributed.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (HideakiImamura) [reply optuna/integration/pytorch_distributed.py:38]: In a distributed optimization setup, even global variables are not shared in memory across processes, right? I think it is equivalent to storing information to properly identify the process group for each `TorchDistributedTrial`.

**Comment by HideakiImamura** [comment]
Let me confirm that one thing. Does not the current PyTorch DDP integration in Optuna's `master` support GPU distributed data parallel training? If it already supports GPU distributed data parallel training, we don't have to change codes. It would be great to share the examples (scripts and environment information) to fail the GPU distributed data parallel training.

**Comment by toshihikoyanase** [comment]
> Does not the current PyTorch DDP integration in Optuna's master support GPU distributed data parallel training?

@reyoung @HideakiImamura I'm sorry for the delayed response.
I confirmed `PytorchDistributedTrial` in the current master supported multi-node configuration.
Could you check the following PR in `optuna/optuna-exampels`, please?

https://github.com/optuna/optuna-examples/pull/150

**Comment by reyoung** [comment]
>>  Does not the current PyTorch DDP integration in Optuna's master support GPU distributed data parallel training?

Yes, you can run GPU DDP with current `Optuna`'s code, BUT:

*  `device` must be set to PytorchDistributedTrial, and `device` is not related to distributed communication at all. They are NOT SAME LEVEL CONCEPTS at all. 
   * The ProcessGroup is decoupled with `device`. Some ProcessGroup has limitation to communicate tensors in some device.
   * However, you cannot just assume this limitation will binding to `device`. Because
      * You can write a ProcessGroup can both support GPU/CPU communication.
*  It must copy between GPU-CPU memory, which is slow and unnecessary.
* It cannot set custom ProcessGroup to handle communication, like pytorch alway can.
  * Some cloud cluster has there own optimized ProcessGroup, like `AWS Segmaker`

It is not common way to set `device` in communication rather than `ProcessGroup`.

I think IT IS A REALLY BAD DESIGN TO USE `DEVICE` RATHER THAN `ProcessGroup` in PytorchDistributedTrial.

**Comment by toshihikoyanase** [comment]
Thank you for your explanation.

The current design of `TorchDistributedTrial` is based on [`ChainerMNStudy`](https://github.com/optuna/optuna/blob/7cb75a060833e7d934fa7164c834d103288603d5/optuna/integration/chainermn.py#L58) that uses the communication paths of weights and gradients to broadcast Optuna's data. So, we intentionally use `device` parameter for `TorchDistributedTrial` to reuse the process group for weights and gradients. And also, the overhead to copy between GPU-CPU is limited, since the Optuna's communication frequency is quite low compared with that for training.

But, as you mentioned, the implementation will be straightforward if we use PyTorch's process group. I'll check the change with `pytorch_distributed_simple.py` in optuna examples.

**Review by HideakiImamura — APPROVED** [review]
LGTM. The followings are follow-ups.

- Fix optuna-examples.
- Reconsider the `_g_pg` variable.
- Add a test case.

**Review by toshihikoyanase — APPROVED** [review]
I confirmed that the change worked with https://github.com/optuna/optuna-examples/pull/150.

Possible follow-up tasks are as follows:
- Replace `_g_pg` with `self._group`
- Always create process group for optuna trials
- Remove `device`
- Add tests for `group`
- Fix docstring for `group`

We'll work on them after merging this PR.

**Comment by toshihikoyanase** [comment]
@reyoung Thank you for your contribution!
